### ISSUE TITLE
Computing Jacobian and Gradient with GradientTape is extremely slow

### ISSUE THREAD

**Issue by melioristic** [tensorflow/tensorflow#52001]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS Linux 7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.6.0
- Python version: 3.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I want to compute the Jacobian and Gradient of my model for a loss function, but the model just throws a warning about retracing and gets stuck

**Describe the expected behavior**
Normal training should occur

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
def compute_loss_theta(tape, parameter, concept, output, x):

    b = x.shape[0]
    in_dim = (x.shape[1], x.shape[2])

    feature_dim = in_dim[0]*in_dim[1]

    J = tape.batch_jacobian(concept, x)
    
    grad_fx = tape.gradient(output, x)
    grad_fx = tf.reshape(grad_fx,shape=(b, feature_dim))
    J = tf.reshape(J, shape=(b, feature_dim, feature_dim))

    parameter = tf.expand_dims(parameter, axis =1)

    loss_theta_matrix = grad_fx - tf.matmul(parameter, J)

    loss_theta = tf.norm(loss_theta_matrix)

    return loss_theta


for i in range(10):
    for x, y in train_dataset:

        with tf.GradientTape(persistent=True) as tape:
            tape.watch(x)
            
            parameter, concept, output = model(x)

            loss_theta = compute_loss_theta(tape, parameter, concept, output , x)

            loss_y = loss_object(y_true=y, y_pred=output)
                
            loss_value = loss_y + eps*loss_theta

        gradients = tape.gradient(loss_value, model.trainable_weights)
        optimizer.apply_gradients(zip(gradients, model.trainable_weights))
```

This function is extremely slow. I want to compute the loss as given in the self-explanatory neural network. Input has a shape of (32, 365, 3) where 32 is the batch size. The loss I want to minimize is Equation 3 of the [paper](https://papers.nips.cc/paper/2018/file/3e9f0fc9b2f89e043bc6233994dfcf76-Paper.pdf). 

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by mohantym** [comment]
Hi @melioristic ,Could you please providing a link  the training_dataset and update the full error stack trace in above template too? It helps expedite the issue.

**Comment by melioristic** [comment]
@mohantym I tried to reproduce this slow behavior. And then looking through the details I found that the Jacobian part is not slow. I was just trying to predict long sequences. You can close this issue. Just on a side note, can you please look through the [notebook](https://colab.research.google.com/drive/1jPPbaxDpkrMSgxS_APtrYGWw1xhbp7lg?usp=sharing) that I have shared here? In the last block can you just see the computation and let me know if this is the fastest way to update the gradients or not. Any suggestion would be really helpful. 
P.S. you can close the issue if you find this fine.

**Comment by mohantym** [comment]
@melioristic  , I was able to remove the warning messages though by changing tabs to put tape.gradient() and opt.apply_gradients() to keep them in context with tf.gradient. Providing [GIST ](https://colab.research.google.com/gist/mohantym/209d13b76a8ef6d50f7057d587991e85/senn_reproducible.ipynb#scrollTo=tS0fJ6emB4GW)for reference.

**Comment by melioristic** [comment]
@mohantym Thanks I think, I will assume that this is the fastest it can work in eager mode. I made this a function and added @tf.function decorator and its speed increased considerably.

**Comment by mohantym** [comment]
Yes ,Agreed! @melioristic ,Checked with `@tf.function(autograph=True) ` earlier  , was able to reduce  existing time  approx 4 times (Around 0.25 after first two batch) . Ok! Feel to Free to close this if it helped .Thanks!
